### PR TITLE
Migrate repr,  deterministic state_dict test to OptimizerInfo

### PR DESCRIPTION
**Pull request by janeyx99**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #116509
* #116500
* __->__ #116496
* #116471

### CODE REVIEW

File: test/test_optim.py [hunk 1]
```diff
@@ -383,6 +385,70 @@ def closure():
...
+            closure = functools.partial(fwd_bwd, optimizer, weight, bias, input)
+
+            # Prime the optimizer
+            for _ in range(10):
+                optimizer.step(closure)
+
+            # Clone the weights and construct a new optimizer for them
+            with torch.no_grad():
+                weight_c = Parameter(weight.clone().detach())
+                bias_c = Parameter(bias.clone().detach())
```
Comment (albanD) [inline test/test_optim.py:426]: nit: detach() does nothing here since you already clone in no_grad mode

**Review by albanD — APPROVED** [review]
Nice

File: test/test_optim.py (hunk 1 above)
Comment (janeyx99) [reply test/test_optim.py:426]: :o
